tests/pf/test_pf_quadtree_inversion_linear.py

```diff
@@ -0,0 +1,242 @@
+            reg.norms = np.c_[0, 0, 0, 0]
+            reg.mref = np.zeros(self.mesh.nC)
+
+            # Data misfit function
+            dmis = data_misfit.L2DataMisfit(simulation=sim, data=data)
+
+            # Add directives to the inversion
+            opt = optimization.ProjectedGNCG(
+                maxIter=15, lower=-1.0, upper=1.0, maxIterLS=5, maxIterCG=5, tolCG=1e-4,
+            )
+
+            invProb = inverse_problem.BaseInvProblem(dmis, reg, opt, beta=beta)
+
+            # Build directives
+            IRLS = directives.Update_IRLS(
+                f_min_change=1e-3,
+                max_irls_iterations=20,
+                beta_tol=1e-1,
+                beta_search=False,
+            )
+            sensitivity_weights = directives.UpdateSensitivityWeights()
+            update_Jacobi = directives.UpdatePreconditioner()
+            inv = inversion.BaseInversion(
+                invProb, directiveList=[IRLS, sensitivity_weights, update_Jacobi]
+            )
+
+            return inv
+
+        # Create grid of points for topography
+        topo_xyz = create_xyz_points(
+            x_range=(-200, 200), y_range=(-200, 200), spacing=50, altitude=0
+        )
+
+        # Create and array of observation points
+        data_xyz = create_xyz_points(
+            x_range=(-100, 100), y_range=(-100, 100), spacing=20, altitude=5
+        )
+
+        # Create the mesh
+        create_mesh(
+            h=[5, 5],
+            padDist=np.ones((2, 2)) * 100,
+            nCpad=[2, 4],
+            topo=topo_xyz,
+            data=data_xyz,
+            min_height=50.0,
+        )
+
+        # Create a density model and generate data,
+        # with a block in a half space
+        self.model = utils.model_builder.addBlock(
+            self.mesh.cell_centers,
+            np.zeros(self.mesh.nC),
+            np.r_[-20, -20],
+            np.r_[20, 20],
+            1.0,
+        )
+
+        # Create reduced identity map. All cells are active in an quadtree
+        self.idenMap = maps.IdentityMap(nP=self.mesh.nC)
+
+        create_gravity_sim(self, block_value=0.3, noise_floor=0.01)
+        self.grav_inv = create_inversion(self, self.grav_sim, self.grav_data, beta=1e3)
+
+        create_magnetics_sim(self, block_value=0.03, noise_floor=3.0)
+        self.mag_inv = create_inversion(self, self.mag_sim, self.mag_data, beta=1e3)
+
+    def test_quadtree_grav_inverse(self):
+
+        # Run the inversion from a zero starting model
+        mrec = self.grav_inv.run(np.zeros(self.mesh.nC))
+
+        # Compute predicted data
+        dpred = self.grav_sim.dpred(self.grav_model)
+
+        # Check models match well enough (allowing for random noise)
+        model_residual = np.linalg.norm(mrec - self.grav_model) / np.linalg.norm(
+            self.grav_model
+        )
+        self.assertAlmostEqual(model_residual, 0.57, delta=0.1)
+
+        # Check data converged to less than 10% of target misfit
+        data_rms = (
```

Comment:
we could have used the `dmis` object here for simplicity if it was saved as `self.dmis`:
`self.dmis(mrec) == 2. * data_rms` here

 but I am ok if left as is.

Reply:
Thanks for the tip @thibaut-kobold!